src/multiplex/multiplex.rs

```diff
@@ -530,3 +727,187 @@ impl<S, T, E> Future for Multiplex<S, T>
         Ok(Async::NotReady)
     }
 }
+
+impl<T: Dispatch> Exchange<T> {
+    fn new(request: Request<T>, deque: FrameDeque<Option<Result<T::BodyOut, T::Error>>>) -> Exchange<T> {
+        Exchange {
+            request: request,
+            responded: false,
+            out_body: None,
+            out_deque: deque,
+            out_is_ready: false,
+            in_body: None,
+        }
+    }
+
+    fn is_inbound(&self) -> bool {
+        match self.request {
+            Request::In => true,
+            Request::Out(_) => false,
+        }
+    }
+
+    fn is_outbound(&self) -> bool {
+        !self.is_inbound()
+    }
+
+    fn is_dispatched(&self) -> bool {
+        match self.request {
+            Request::Out(Some(_)) => false,
+            _ => true,
+        }
+    }
+
+    /// Returns true if the exchange is complete
+    fn is_complete(&self) -> bool {
+        // The exchange is completed if the response has been seen and bodies
+        // in both directions are fully flushed
+        self.responded && self.out_body.is_none() && self.in_body.is_none()
+    }
+
+    /// Takes the buffered out request out of the value and returns it
+    fn take_buffered_out_request(&mut self) -> Option<Message<T::Out, Body<T::BodyOut, T::Error>>> {
+        match self.request {
+            Request::Out(ref mut request) => request.take(),
+            _ => None,
+        }
+    }
+
+    fn send_out_chunk(&mut self, chunk: Result<Option<T::BodyOut>, T::Error>) {
+        // Reverse Result & Option
+        let chunk = match chunk {
+            Ok(Some(v)) => Some(Ok(v)),
+            Ok(None) => None,
+            Err(e) => Some(Err(e)),
+        };
+
+        // Get a reference to the sender
+        {
+            let sender = match self.out_body {
+                Some(ref mut v) => v,
+                _ =>  {
+                    return;
+                }
+            };
+
+            if self.out_is_ready {
```

Comment:
Hm ok so looking more above, I got pretty confused by `out_is_ready` unfortunately. Maybe this method could internally call flush and detect that the flush didn't go all the way and buffer + return? I've found it helpful at least to avoid state flags and go to the "source of truth" (the underlying channel) as much as possible at least. Not sure if that has a perf impact though.